### ISSUE TITLE
mini.trailspace adds trailing space highlighting in snacks.nvim dashboard

### ISSUE THREAD

**Issue by ronisbr** [echasnovski/mini.nvim#1395]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)
- [X] I have updated 'mini.nvim' to latest version

### Module(s)

mini.trailspace

### Description

If you enable mini.trailspace and snacks.nvim dashboard, you will see trailspace highlighting until an highlight update happens.

### Neovim version

0.10.1

### Steps to reproduce

1. Install snacks.nvim (with dashboard) and mini.trailspace.
2. Open neovim.

### Actual behavior

When Noevim is opened without an input file, I see this result:

<img width="1153" alt="Screenshot 2024-12-12 at 11 55 31" src="https://github.com/user-attachments/assets/89c4c1d3-f92d-4d94-ae11-c63b702e2c04" />

If I open a new buffer, like lazy configuration window, the highlight gets updated and everything is correct:

<img width="1153" alt="Screenshot 2024-12-12 at 11 56 02" src="https://github.com/user-attachments/assets/f9ec54a9-6c85-48f0-8b23-221354d7ebdb" />

The problem does not happen if mini.trailspace is disabled.

**Comment by echasnovski** [comment]
'mini.trailspace' and most other 'mini.nvim' modules can be disabled per buffer by setting [an appropriate buffer-local variable](https://github.com/echasnovski/mini.nvim/blob/d084df4d6bf58abcc768b9bb326241ff0c502793/doc/mini-trailspace.txt#L35-L43). So in this case it is the matter of setting `vim.b.minitrailspace_disable` before the dashboard buffer dashboard buffer is entered.

**Comment by ronisbr** [comment]
I see! Thanks!

**Comment by ronisbr** [comment]
For reference, I could solve the problem by enabling mini.trailspace only after a new buffer is created:

```lua
    config = function(_, opts)
      require("mini.trailspace").setup(opts)

      -- HACK: We need to disabel the mini.trailspace and enable when a new buffer is
      -- created to avoid interference with the dashboard snacks.nvim. See:
      --
      --  https://github.com/echasnovski/mini.nvim/issues/1395
      vim.g.minitrailspace_disable = true

      vim.api.nvim_create_autocmd("BufNew", {
        callback = function()
          vim.g.minitrailspace_disable = false
        end
      })
```

It is probably not the best solution, but it works.

**Comment by echasnovski** [comment]
I'd suggest disabling locally to dashboard's buffer with `vim.b.minitrailspace_disable = true` set only in necessary buffer(s). Here is [a recipe for common case of disabling based on filetype](https://github.com/echasnovski/mini.nvim/blob/e78a98ac2f984fb30e31c016f97d57e2d60a3ee7/doc/mini.txt#L339-L341). The filetype [seems to be `snacks_dashboad`](https://github.com/folke/snacks.nvim/blob/52086010d3f72949a30c45c2cc4faa0b0286d097/lua/snacks/dashboard.lua#L156).

**Comment by ronisbr** [comment]
Thanks @echasnovski !

This was my very first attempt, but, for some reason, the autocmd was not triggered. I think it has to be something related to the vim initialization that I do not fully understand.

**Comment by echasnovski** [comment]
Alternatively, there seems to be a 'SnacksDashboardOpened' event. Maybe something like this:

```lua
local f = function(args) vim.b[args.buf].minitrailspace_disable = true end
vim.api.nvim_create_autocmd('User', { pattern = 'SnacksDashboardOpened', callback = f })
```

**Comment by ronisbr** [comment]
This suggestion worked perfectly! Thanks @echasnovski !

**Comment by SupremeTechnopriest** [comment]
It works but it still flashes red on start. Would be nice to have a `disabledFileTypes` setting or similar to avoid these race conditions.

**Comment by echasnovski** [comment]
> It works but it still flashes red on start. Would be nice to have a `disabledFileTypes` setting or similar to avoid these race conditions.

The potential `disabledFileTypes` is essentially the same as [suggested recipe for disabling per filetype](https://github.com/echasnovski/mini.nvim/blob/e78a98ac2f984fb30e31c016f97d57e2d60a3ee7/doc/mini.txt#L339-L341). There should be no flashing if buffer-local variable is set during startup (i.e. not lazy loaded).

**Comment by SupremeTechnopriest** [comment]
I have both mini and snacks set to `lazy=false`. mini has priority `9999` and snacks has priority `999`.  I have the following snippet in mini's `config` function:

```lua
		local disabled_line_decoration = {
			"snacks_dashboard",
			"oil",
			"notify",
			"snacks_notif",
			"lazy",
			"mason",
			"noice",
		}

		local disabled_animation = {
			"snacks_dashboard",
			"notify",
			"noice",
			"snacks_notif",
			"oil",
		}

		vim.api.nvim_create_autocmd("Filetype", {
			callback = function(args)
				local ft = vim.bo[args.buf].filetype
				if not vim.tbl_contains(disabled_line_decoration, ft) then
					vim.b[args.buf].minitrailspace_disable = true
				end
				if vim.tbl_contains(disabled_animation, ft) then
					vim.b[args.buf].minianimate_disable = true
				end
			end,
		})

		vim.api.nvim_create_autocmd("User", {
			pattern = "SnacksDashboardOpened",
			callback = function(args)
				vim.b[args.buf].minitrailspace_disable = true
			end,
		})
```

Am I missing something? I still see trailspace flashing before its disabled.

**Comment by echasnovski** [comment]
> I have both mini and snacks set to `lazy=false`. mini has priority `9999` and snacks has priority `999`.
> ...
> Am I missing something? I still see trailspace flashing before its disabled.

Sorry, I am not proficient with setting up 'lazy.nvim'. Here is the config that works without flashing using 'mini.nvim' and 'mini.deps':

```lua
local path_package = vim.fn.stdpath('data') .. '/site/'
local mini_path = path_package .. 'pack/deps/start/mini.nvim'
if not vim.loop.fs_stat(mini_path) then
  vim.cmd('echo "Installing `mini.nvim`" | redraw')
  local clone_cmd = { 'git', 'clone', '--filter=blob:none', 'https://github.com/echasnovski/mini.nvim', mini_path }
  vim.fn.system(clone_cmd)
  vim.cmd('packadd mini.nvim | helptags ALL')
  vim.cmd('echo "Installed `mini.nvim`" | redraw')
end

-- Make sure 'mini.nvim' is available
vim.cmd("packadd mini.nvim")

require("mini.deps").setup()

require("mini.trailspace").setup()
vim.api.nvim_create_autocmd("User", {
	pattern = "SnacksDashboardOpened",
	callback = function(args)
		vim.b[args.buf].minitrailspace_disable = true
	end,
})

MiniDeps.add("folke/snacks.nvim")
require("snacks").setup({
	dashboard = {
		enable = true,
		sections = {
			{ section = "header" },
			{ section = "keys", gap = 1, padding = 1 },
		},
	},
})
```

**Comment by SupremeTechnopriest** [comment]
Gotcha, thanks @echasnovski.  If I figure out how to eliminate the flash with lazy I will post it here for others who come across it. Otherwise I can live with the flash.

**Comment by BYT0723** [comment]
getting this method works [snacks.nvim #1220](https://github.com/folke/snacks.nvim/issues/1220)
```lua
  {
    'echasnovski/mini.trailspace',
    event = 'BufReadPost',
    opts = {},
  },
```
Lazy loading trailspace via `BufReadPost`

**Comment by BYT0723** [comment]
This configuration works. [snacks.nvim #1220](https://github.com/folke/snacks.nvim/issues/1220#issuecomment-2661542968)
```lua
{
  'echasnovski/mini.trailspace',
  event = 'BufReadPost',
  opts = {},
},
```
lazy loading trailspace via `BufReadPost` event
